backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.cpp

```diff
@@ -0,0 +1,663 @@
+/// with name "(tmp", IR::Expression, IR::LAnd, IR::Expression] The result will be an IR::Expression
+/// equal to !IR::Expression || (IR::Expression && IR::Expression)
+const IR::Expression* makeSingleExpr(std::vector<const IR::Expression*> input) {
+    const IR::Expression* expr = nullptr;
+    for (uint64_t i = 0; i < input.size(); i++) {
+        if (const auto* lOr = input[i]->to<IR::LOr>()) {
+            if (lOr->right->toString() == "(tmp") {
+                if (i + 1 == input.size()) {
+                    break;
+                }
+                std::vector<const IR::Expression*> tmp;
+                uint64_t idx = i;
+                i++;
+                if (i + 1 == input.size()) {
+                    expr = new IR::LOr(input[idx - 1], input[i]);
+                    break;
+                }
+                while (i < input.size()) {
+                    tmp.push_back(input[i]);
+                    i++;
+                }
+                if (expr == nullptr) {
+                    expr = new IR::LOr(input[idx - 1], makeSingleExpr(tmp));
+                } else {
+                    expr = new IR::LOr(expr, makeSingleExpr(tmp));
+                }
+                break;
+            }
+            {
+                if (i + 1 == input.size()) {
+                    break;
+                }
+                if (expr == nullptr) {
+                    expr = new IR::LOr(input[i - 1], input[i + 1]);
+                } else {
+                    expr = new IR::LOr(expr, input[i + 1]);
+                }
+            }
+        }
+        if (input[i]->is<IR::LAnd>()) {
+            if (i + 1 == input.size()) {
+                break;
+            }
+            if (expr == nullptr) {
+                expr = new IR::LAnd(input[i - 1], input[i + 1]);
+            } else {
+                expr = new IR::LAnd(expr, input[i + 1]);
+            }
+        }
+    }
+
+    return expr;
+}
+
+/// Determines the token type according to the table key and generates a zombie constant for it.
+const IR::Expression* makeConstant(Token input, const IR::Vector<IR::KeyElement>& keyElements,
+                                   const IR::Type* leftType) {
+    const IR::Type_Base* type = nullptr;
+    const IR::Expression* result = nullptr;
+    auto inputStr = input.lexeme();
+    if (input.is(Token::Kind::Text)) {
+        for (const auto* key : keyElements) {
+            cstring annotationName;
+            if (const auto* annotation = key->getAnnotation("name")) {
+                if (!annotation->body.empty()) {
+                    annotationName = annotation->body[0]->text;
+                } else if (!annotation->expr.empty()) {
+                    annotationName = annotation->expr[0]->toString();
+                }
+            }
+            BUG_CHECK(annotationName.size() > 0, "Key does not have a name annotation.");
+            cstring separator = "\"";
```

Comment:
Can you please add a comment explaining why this is needed or what it is for?

Reply:
I don't know for what reason, but when I take `toString` from expr in annotation , it returns the string `"keyname"` to me and not `keyname`. Due to the fact that `"` remains in the string, the algorithm cannot compare the names